<a href="https://colab.research.google.com/github/vinigm/Portfolio/blob/main/3_SalesAnalysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Sales Analysis**
---

[VÍDEO](https://www.youtube.com/watch?v=eMOA1pPVUc4&list=LL&index=1&t=340s)

Esse projeto consiste na análise de planilhas mensais de faturamento de uma loja. A modelagem se faz presente para uma melhor compreensão de direcionamento estratégico de custos.

Infelizmente as tabelas não possuem as marcas dos produtos, mas temos discriminados os nomes desses, bem como datas das ordens de compras e locais aos quais foram destinados.

## **Importação de bibliotecas e dados**

In [42]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import datetime as dt

Os dados são referentes às vendas de produtos ao longo de um ano. Os datasets estão divididos por mês.

In [43]:
datajan = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/Pandas-Data-Science-Tasks/master/SalesAnalysis/Sales_Data/Sales_January_2019.csv') #January
datafeb = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/Pandas-Data-Science-Tasks/master/SalesAnalysis/Sales_Data/Sales_February_2019.csv') #February
datamarch = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/Pandas-Data-Science-Tasks/master/SalesAnalysis/Sales_Data/Sales_March_2019.csv') #March
dataabril = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/Pandas-Data-Science-Tasks/master/SalesAnalysis/Sales_Data/Sales_April_2019.csv') #Abril
datamay = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/Pandas-Data-Science-Tasks/master/SalesAnalysis/Sales_Data/Sales_May_2019.csv') #May
datajun = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/Pandas-Data-Science-Tasks/master/SalesAnalysis/Sales_Data/Sales_June_2019.csv') #June
datajul = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/Pandas-Data-Science-Tasks/master/SalesAnalysis/Sales_Data/Sales_July_2019.csv') #July
dataaug = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/Pandas-Data-Science-Tasks/master/SalesAnalysis/Sales_Data/Sales_August_2019.csv') #August
datasep = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/Pandas-Data-Science-Tasks/master/SalesAnalysis/Sales_Data/Sales_September_2019.csv') #September
dataoc = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/Pandas-Data-Science-Tasks/master/SalesAnalysis/Sales_Data/Sales_October_2019.csv') #October
datanov = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/Pandas-Data-Science-Tasks/master/SalesAnalysis/Sales_Data/Sales_November_2019.csv') #November
datadez = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/Pandas-Data-Science-Tasks/master/SalesAnalysis/Sales_Data/Sales_December_2019.csv') #December

**Transformação em DataFrames** de todas as tabelas individuais.

In [44]:
dfjan = pd.DataFrame(datajan)
dffeb = pd.DataFrame(datafeb)
dfmarch = pd.DataFrame(datamarch)
dfabril = pd.DataFrame(dataabril)
dfmay = pd.DataFrame(datamay)
dfjun = pd.DataFrame(datajun)
dfjul = pd.DataFrame(datajul)
dfaug = pd.DataFrame(dataaug)
dfsep = pd.DataFrame(datasep)
dfoc = pd.DataFrame(dataoc)
dfnov = pd.DataFrame(datanov)
dfdez = pd.DataFrame(datadez)

**Concatenação** das tabelas para melhor trabalhabilidade.

In [45]:
df = pd.concat([dfjan, dffeb, dfmarch, dfmay, dfjun, dfjul, dfaug, dfsep, dfoc, dfnov, dfdez])
df.head()

Order ID                   Product Quantity Ordered Price Each  \
0   141234                    iPhone                1        700   
1   141235  Lightning Charging Cable                1      14.95   
2   141236          Wired Headphones                2      11.99   
3   141237          27in FHD Monitor                1     149.99   
4   141238          Wired Headphones                1      11.99   

       Order Date                       Purchase Address  
0  01/22/19 21:25        944 Walnut St, Boston, MA 02215  
1  01/28/19 14:15       185 Maple St, Portland, OR 97035  
2  01/17/19 13:33  538 Adams St, San Francisco, CA 94016  
3  01/05/19 20:33     738 10th St, Los Angeles, CA 90001  
4  01/25/19 11:59          387 10th St, Austin, TX 73301

## **Exploração e Limpeza dos Dados**

**Renomeando** colunas para melhor trabalhabilidade.

In [46]:
df.columns = df.columns.str.replace(' ','_')

df.rename(columns={"Order_ID": "order_id",
                     "Product": "product",
                     "Quantity_Ordered": "quantity_ordered",
                     "Price_Each": "price_each",
                     "Order_Date": "order_date",
                     "Purchase_Address": "purchase_address"
                     }, inplace=True)

In [47]:
df.head()

order_id                   product quantity_ordered price_each  \
0   141234                    iPhone                1        700   
1   141235  Lightning Charging Cable                1      14.95   
2   141236          Wired Headphones                2      11.99   
3   141237          27in FHD Monitor                1     149.99   
4   141238          Wired Headphones                1      11.99   

       order_date                       purchase_address  
0  01/22/19 21:25        944 Walnut St, Boston, MA 02215  
1  01/28/19 14:15       185 Maple St, Portland, OR 97035  
2  01/17/19 13:33  538 Adams St, San Francisco, CA 94016  
3  01/05/19 20:33     738 10th St, Los Angeles, CA 90001  
4  01/25/19 11:59          387 10th St, Austin, TX 73301

In [48]:
df.shape

(168467, 6)

In [49]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 168467 entries, 0 to 25116
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   order_id          167981 non-null  object
 1   product           167981 non-null  object
 2   quantity_ordered  167981 non-null  object
 3   price_each        167981 non-null  object
 4   order_date        167981 non-null  object
 5   purchase_address  167981 non-null  object
dtypes: object(6)
memory usage: 9.0+ MB


**Verificação** de valores não nulos

In [50]:
df.isnull().sum()      # Necessário um tratamento nos dados com valores nulos.

order_id            486
product             486
quantity_ordered    486
price_each          486
order_date          486
purchase_address    486
dtype: int64

In [51]:
# Porcentagem de dados faltantes no DF (em %)

faltantes = (df.isnull().sum() / len(df['order_id']))*100
print(faltantes)

order_id            0.288484
product             0.288484
quantity_ordered    0.288484
price_each          0.288484
order_date          0.288484
purchase_address    0.288484
dtype: float64


In [52]:
nan_df = df[df.isna().any(axis=1)]
nan_df.head()

order_id product quantity_ordered price_each order_date purchase_address
664       NaN     NaN              NaN        NaN        NaN              NaN
678       NaN     NaN              NaN        NaN        NaN              NaN
797       NaN     NaN              NaN        NaN        NaN              NaN
876       NaN     NaN              NaN        NaN        NaN              NaN
1299      NaN     NaN              NaN        NaN        NaN              NaN

In [53]:
df2 = df.dropna()

In [54]:
df2.shape

(167981, 6)

In [55]:
# Não foi possível transformar a coluna "month" para "integer" pois o erro acusava que existia algum dado com valor de "Or"
# Assim, apenas para fins de exploração se faz o DF temporário com uma coluna apenas com meses com valor igual à "Or".

temp_df = df2[df2['order_date'].str[0:2] == 'Or']
temp_df.head()

order_id  product  quantity_ordered  price_each  order_date  \
1073  Order ID  Product  Quantity Ordered  Price Each  Order Date   
1102  Order ID  Product  Quantity Ordered  Price Each  Order Date   
1194  Order ID  Product  Quantity Ordered  Price Each  Order Date   
1897  Order ID  Product  Quantity Ordered  Price Each  Order Date   
2463  Order ID  Product  Quantity Ordered  Price Each  Order Date   

      purchase_address  
1073  Purchase Address  
1102  Purchase Address  
1194  Purchase Address  
1897  Purchase Address  
2463  Purchase Address

In [56]:
len(temp_df)

320

Como temos 320 linhas apenas, isso representa algo em torno de 0.28% do nosso dataframe. Desprezivel em relação ao valor global, assim, retira-se essas linhas para dar sequencia na análise.

In [57]:
df2 = df2[df2['order_date'].str[0:2] != 'Or']

In [58]:
len(df2)

167661

- len(df) = 168.467
- len(df2) = 167.981 (Depois de retirar os valores nulos)
- len(df2) = 167.661 (Depois de retirar os valores nulos e "Or's"

## **Qual o mês que vendeu mais?**

**Criação** da coluna mês e conversão dos dados para "int32"

In [59]:
df2['month'] = df2['order_date'].str[0:2]
df2['month'] = df2['month'].astype('int32')

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [60]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 167661 entries, 0 to 25116
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   order_id          167661 non-null  object
 1   product           167661 non-null  object
 2   quantity_ordered  167661 non-null  object
 3   price_each        167661 non-null  object
 4   order_date        167661 non-null  object
 5   purchase_address  167661 non-null  object
 6   month             167661 non-null  int32 
dtypes: int32(1), object(6)
memory usage: 9.6+ MB


In [63]:
df2.head()

order_id                   product quantity_ordered price_each  \
0   141234                    iPhone                1        700   
1   141235  Lightning Charging Cable                1      14.95   
2   141236          Wired Headphones                2      11.99   
3   141237          27in FHD Monitor                1     149.99   
4   141238          Wired Headphones                1      11.99   

       order_date                       purchase_address  month  
0  01/22/19 21:25        944 Walnut St, Boston, MA 02215      1  
1  01/28/19 14:15       185 Maple St, Portland, OR 97035      1  
2  01/17/19 13:33  538 Adams St, San Francisco, CA 94016      1  
3  01/05/19 20:33     738 10th St, Los Angeles, CA 90001      1  
4  01/25/19 11:59          387 10th St, Austin, TX 73301      1

In [ ]:
df2['order_sale'] = df['quantity_ordered']*df['price_each']